clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -183,22 +183,33 @@ public RecordAppendResult append(TopicPartition tp,
             ByteBuffer buffer = free.allocate(size, maxTimeToBlock);
             synchronized (dq) {
                 // Need to check if producer is closed again after grabbing the dequeue lock.
-                if (closed)
+                if (closed) {
+                    free.deallocate(buffer);
                     throw new IllegalStateException("Cannot send after the producer is closed.");
+                }
+
+                RecordAppendResult appendResult = null;
+                try {
+                    appendResult = tryAppend(timestamp, key, value, callback, dq);
+                } catch (Exception e) {
```

Comment:
@satishd I re-read your code and I see that you are only de-allocating on exceptions instead of using a finally like in my suggestion. They both work (in my suggestion, we null the outer variable once we have used it to prevent it from being freed). One difference is that in your approach, we could end up calling `free.deallocate` twice if the first invocation throws an exception (which should not happen normally).